styx-scheduler-service/src/main/java/com/spotify/styx/state/PersistentStateManager.java

```diff
@@ -353,7 +359,7 @@ private void verifyCounter(Event event, long expectedCounter, RunState currentRu
                              + expectedCounter  + " but current counter is "
                              + currentCounter + ". Discarding event " + event;
       log.debug(message);
-      throw new StaleEventException(message);
+      throw new StateTransitionConflictException(message);
```

Comment:
Is it true this can only be caused by concurrent transition?

Reply:
It could also be caused by retries without fetching the latest counter. But anyway we are ignoring this exception.

Reply:
Yes, this could mask incorrect invocation of `receive` with an _incorrect_ counter. And we want to ignore the (hopefully) majority of cases when `receive` was invoked with a correct (although possibly stale due to concurrent transitions) counter. Here's a few ways to address that:

1. Stop using counters and wrap all pairs of read + update run-state operations into a single transaction. This seems "right" to me, and is what one would normally do when implementing transactional state processing. Maybe the reason for us using counters has become "stale" (pun intended).

2. Minimize the number of `receive` call-sites by only having them in the PSM tick. This entails making `OutputHandlers` return the events they want to emit instead of calling the state manager directly.

These are a bit outside the scope of this PR though.